addons/image/freeimage.c

```diff
@@ -0,0 +1,168 @@
+/* Allegro wrapper routines for FreeImage
+ * by Karthik Kumar Viswanathan <karthikkumar@gmail.com>.
+ */
+
+#include <FreeImage.h>
+
+#include "allegro5/allegro.h"
+#include "allegro5/allegro_image.h"
+#include "allegro5/internal/aintern_exitfunc.h"
+#include "allegro5/internal/aintern_image.h"
+
+#include "iio.h"
+
+ALLEGRO_DEBUG_CHANNEL("image")
+
+static bool freeimage_initialized = false;
+
+bool _al_init_fi(void)
+{
+   if(freeimage_initialized)
+       return true;
+   FreeImage_Initialise(FALSE);
+   _al_add_exit_func(_al_shutdown_fi, "_al_shutdown_fi");
+   return true;
+}
+
+void _al_shutdown_fi(void)
+{
+   if(!freeimage_initialized)
+       return;
+   FreeImage_DeInitialise();
+   freeimage_initialized = false;
+}
+
+static ALLEGRO_BITMAP *_al_fi_to_al_bitmap(FIBITMAP *fib) {
+   int width = 0, height = 0;
+   ALLEGRO_BITMAP *bitmap = NULL;
+   width = FreeImage_GetWidth(fib);
+   height = FreeImage_GetHeight(fib);
+   bitmap = al_create_bitmap(width, height);
+   if(bitmap) {
+      ALLEGRO_LOCKED_REGION *a_lock = al_lock_bitmap(bitmap,
+         ALLEGRO_PIXEL_FORMAT_ARGB_8888, ALLEGRO_LOCK_WRITEONLY);
+      if (a_lock) {
+         unsigned char *out = (unsigned char *)a_lock->data;
+         int out_inc = a_lock->pitch - (width*4);
+         for(int j=height - 1; j > -1; --j) {
+            for(int i=0; i < width; ++i) {
+               RGBQUAD color;
+               if(FreeImage_GetPixelColor(fib, i, j, &color) == FALSE) {
+                  ALLEGRO_ERROR("Unable to get pixel data at %d,%d\n", i , j);
+               }
+               *out++ = (unsigned char) color.rgbBlue;
+               *out++ = (unsigned char) color.rgbGreen;
+               *out++ = (unsigned char) color.rgbRed;
+               *out++ = (unsigned char) color.rgbReserved;
+            }
+            out += out_inc;
+         }
+         al_unlock_bitmap(bitmap);
+      }
+   }
+   return bitmap;
+}
+
+ALLEGRO_BITMAP *_al_load_fi_bitmap(const char *filename, int flags)
+{
+   FIBITMAP *fib = NULL;
+   ALLEGRO_BITMAP *bitmap = NULL;
+   FREE_IMAGE_FORMAT fif = FIF_UNKNOWN;
+
+   ASSERT(filename);
+   ASSERT(freeimage_initialized == true);
+
+   fif = FreeImage_GetFIFFromFilename(filename);
+   if(fif == FIF_UNKNOWN)
+      fif = FreeImage_GetFileType(filename, 0);
+   if(fif == FIF_UNKNOWN)
+      return bitmap;
```

Comment:
`return NULL` would be clearer. Also, could you add a `ALLEGRO_WARN("Could not determine the file type for '%s'", filename)`

Reply:
Fixed.